tarfs.go

```diff
@@ -0,0 +1,251 @@
+package tarfs
+
+import (
+	"archive/tar"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/getlantern/golog"
+)
+
+var (
+	log            = golog.LoggerFor("tarfs")
+	fileTimestamp  = time.Now()
+	emptyFileInfos = []os.FileInfo{}
+)
+
+// FileSystem is a tarfs filesystem. It exposes a Get method for accessing
+// resources by path. It also implements http.FileSystem for use with
+// http.FileServer.
+type FileSystem struct {
+	files map[string][]byte
+	local string
+}
+
+// New creates a new FileSystem using the given in-memory tar data. If local is
+// a non-empty string, the resulting FileSystem will first look for resources in
+// the local file system before returning an embedded resource.
+func New(tarData []byte, local string) (*FileSystem, error) {
+	if local != "" {
+		_, err := os.Stat(local)
+		if err != nil {
+			if os.IsNotExist(err) {
+				log.Debugf("Local dir %v does not exist, not using\n", local)
+				local = ""
+			} else {
+				log.Errorf("Unable to stat local dir %v: %v\n", local, err)
+				os.Exit(5)
+			}
+		} else {
+			log.Debugf("Using local filesystem at %v", local)
+		}
+	}
+
+	fs := &FileSystem{make(map[string][]byte, 0), local}
+
+	// Read the tar data and index it into a map
+	br := &trackingreader{bytes.NewReader(tarData), 0}
+	for {
+		// We construct a new tar reader every time because each loop advances
+		// the underlying trackingreader to the next tar header and the existing
+		// tar reader doesn't know it.
+		tr := tar.NewReader(br)
+
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			log.Trace("Reached end of tar archive")
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Unable to read next tar header: %v", err)
+		}
+
+		// Create a slice to the tar data started after the header and
+		// containing the specified size of data. We don't use tr.Read() so that
+		// we can avoid copying.
+		end := br.pos + hdr.Size
+		fs.files[hdr.Name] = tarData[br.pos:end]
+
+		// Advance to the next tar header. Note that we round up to the next
+		// multiple of 512 because tar files contain 512 byte blocks that are 0
+		// padded if the actual data doesn't align with the 512 byte boundaries.
+		err = br.AdvanceTo(int64(math.Ceil(float64(end)/512)) * 512)
```

Comment:
Do we need to make a test file larger than 512 bytes to cover the calculation, which is very important and subtle?